runtime/lua/vim/path.lua

```diff
@@ -0,0 +1,59 @@
+local M = {}
+
+--- Ascend the filesystem tree starting at {start}, calling {func} for each parent directory.
+---
+--- Example:
+--- <pre>
+--- local curdir = vim.fn.fnamemodify(vim.api.nvim_buf_get_name(0), ":p:h")
+--- local gitdir = vim.path.ascend(curdir, function(dir)
+---   -- Stop searching in $HOME
+---   if curdir == vim.env.HOME then
+---     return false
+---   end
+---
+---   if vim.fn.isdirectory(curdir .. "/.git") == 1 then
+---     return true
+---   end
+--- end)
+---
+--- if gitdir then
+---   print("Found git directory at " .. gitdir)
+--- else
+---   print("Not in a git repository")
+--- end
+---</pre>
+---
+---@param start (string) Directory to start from
+---@param func (function) A function called for each directory. The function is passed the
+---            full path of the directory as its single argument. If this function returns non-nil
+---            then walking will end.
+---@return (string|nil) If {func} returns true for a given directory, then that directory is the
+---         return value of this function. Otherwise, if {func} returns false or if the root of the
+---         filesystem is reached, this function returns nil.
+function M.ascend(start, func)
+  vim.validate({
+    start = { start, 's' },
+    func = { func, 'f' },
+  })
+
+  local dir = vim.fn.fnamemodify(start, ':p')
```

Comment:
Might be worth marking that this can't be run in api-fast.

Reply:
Could also us pl.path.absolute & pl.path.parent? (or introduce those in `vim.path`).